File: AOC_2023/Day 7/7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
	"sort"
)

func initialize(filename string) (map[string]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	hands := make(map[string]int)
	for scanner.Scan() {
		camelCard := scanner.Text()
		hand := camelCard[:strings.Index(camelCard, " ")]
		card, _ := strconv.Atoi(camelCard[strings.Index(camelCard, " ")+1:])
		hands[hand] = card
	}
	return hands,nil
}

func order(hand string, newOrder bool) int {
	count := map[string]int{}
	for i := 0; i < len(hand); i++ {
		count[string(hand[i])]++
	}
	if !newOrder || count["J"] == 0 {
		if len(count) == 1 {
			return 7
		} else if len(count) == 2 {
			for _, v := range count {
				if v == 4 {
					return 6
				}
			}
			return 5
		} else if len(count) == 3 {
			for _, v := range count {
				if v == 3 {
					return 4
				}
			}
			return 3
		} else if len(count) == 4 {
			return 2
		}
		return 1
	}
	if len(count) == 1 || len(count) == 2 {
		return 7
	}
	if len(count) == 3 {
		for _, v := range count {
			if v == 3 {
				return 6
			}
		}
		if count["J"] == 2 {
			return 6
		}
		return 5
	}
	if len(count) == 4 {
		return 4
	}
	return 2
}

func part1(hands map[string]int) {
	result := 0
	keys := make([]string, 0, len(hands))
	for k := range hands {
		keys = append(keys, k)
	}
	ordering := map[rune]int{'A':14, 'K':13, 'Q':12, 'J':11, 'T':10, '9':9, '8':8, '7':7, '6':6, '5':5, '4':4, '3':3, '2':2}
	sort.Slice(keys, func(i, j int) bool {
		order1 := order(keys[i], false)
		order2 := order(keys[j], false)
		if order1 != order2 {
			return order1 < order2
		}
		for k := 0; k < len(keys[i]); k++ {
			if ordering[rune(keys[i][k])] != ordering[rune(keys[j][k])] {
				return ordering[rune(keys[i][k])] < ordering[rune(keys[j][k])]
			}
		}
		return false
	})
	for i := 0; i < len(keys); i++ {
		result += hands[keys[i]] * (i+1)
	}
	fmt.Println("Part 1:", result)
}

func part2(hands map[string]int) {
	result := 0
	keys := make([]string, 0, len(hands))
	for k := range hands {
		keys = append(keys, k)
	}
	ordering := map[rune]int{'A':14, 'K':13, 'Q':12, 'T': 10, '9':9, '8':8, '7':7, '6':6, '5':5, '4':4, '3':3, '2':2, 'J':1}
	sort.Slice(keys, func(i, j int) bool {
		order1 := order(keys[i], true)
		order2 := order(keys[j], true)
		if order1 != order2 {
			return order1 < order2
		}
		for k := 0; k < len(keys[i]); k++ {
			if ordering[rune(keys[i][k])] != ordering[rune(keys[j][k])] {
				return ordering[rune(keys[i][k])] < ordering[rune(keys[j][k])]
			}
		}
		return false
	})
	for i := 0; i < len(keys); i++ {
		result += hands[keys[i]] * (i+1)
	}
	fmt.Println("Part 2:", result)
}

func main() {
	filename := "input.txt"
	hands, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(hands)
	part2(hands)
}

File: AOC_2023/Day 7/7.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

// Ordering
// Five of a kind
// Four of a kind
// Full house
// Three of a kind
// Two pairs
// One pair
// High card
int ordering(string s, bool newOrder){
    unordered_map<char, int> m;
    for (int i = 0; i < s.length(); i++)
    {
        m[s[i]]++;
    }
    if(!newOrder || m.count('J') == 0){
        // Five of a kind
        if(m.size() == 1) return 7;
        // Four of a kind or Full house
        if(m.size() == 2){
            for (auto i : m)
            {
                if(i.second == 4) return 6;
                if(i.second == 3) return 5;
            }
        }
        // Three of a kind or Two pairs
        if(m.size() == 3){
            for (auto i : m)
            {
                if(i.second == 3) return 4;
                if(i.second == 2) return 3;
            }
        }
        // One pair
        if(m.size() == 4) return 2;
        // High card
        return 1;
    }
    int cnt = m['J'];
    // 5 => five of a kind
    // 4,1 / 3,2 => can be made into fice of A KIND
    if(m.size() == 1 || m.size() == 2) return 7;
    // 3,1,1 => four of a kind
    // 2,2,1 => full house/four of a kind
    if(m.size() == 3){
        for (auto i : m)
        {
            if(i.second == 3) return 6;
            if(i.second == 2) {
                if(cnt == 1) return 5;
                return 6;
            }
        }
    }
    // 2,1,1,1 => three of a kind / two pair but 2 pair always better
    if(m.size() == 4){
        return 4;
    }
    // 1,1,1,1,1 => high card => one pair 
    return 2;
}

// card order : A, K, Q, J, T, 9, 8, 7, 6, 5, 4, 3, or 2
int cardOrder(char c, bool newOrder){
    if(c == 'A') return 14;
    if(c == 'K') return 13;
    if(c == 'Q') return 12;
    if(c == 'J'){
        if(newOrder) return 1;
        return 11;
    }
    if(c == 'T') return 10;
    return c - '0';
}

bool cmp(pair<string, int> a, pair<string, int> b){
    if(ordering(a.first, false) != ordering(b.first, false)){
        return ordering(a.first, false) < ordering(b.first, false);
    }
    else{
        for(int i = 0; i < a.first.length(); i++){
            if(cardOrder(a.first[i],false) != cardOrder(b.first[i], false))
                return cardOrder(a.first[i], false) < cardOrder(b.first[i], false);
        }
    }
    return a.second < b.second;
}

bool cmp2(pair<string, int> a, pair<string, int> b){
    if(ordering(a.first, true) != ordering(b.first, true)){
        return ordering(a.first, true) < ordering(b.first, true);
    }
    else{
        for(int i = 0; i < a.first.length(); i++){
            if(cardOrder(a.first[i],true) != cardOrder(b.first[i], true))
                return cardOrder(a.first[i], true) < cardOrder(b.first[i], true);
        }
    }
    return a.second < b.second;
}

class Solution
{
    public:
    vector<pair<string, int>> hands;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            hands.push_back(make_pair(line.substr(0, line.find(" ")), stoi(line.substr(line.find(" ")+1))));   
        }
    }
    void part1(){
        long long result = 0;
        sort(hands.begin(), hands.end(), cmp);
        for (int i = 0; i < hands.size(); i++)
        {
            result += hands[i].second * (i+1);
        }
        cout<<"Part 1: "<<result<<"\n";
    }
    void part2(){
        long long result = 0;
        sort(hands.begin(), hands.end(), cmp2);
        for (int i = 0; i < hands.size(); i++)
        {
            result += hands[i].second * (i+1);
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}